packages/fxa-auth-server/lib/routes/subscriptions/paypal-notifications.ts

```diff
@@ -11,7 +13,60 @@ import { AuthLogger, AuthRequest } from '../../types';
 import { PayPalHandler } from './paypal';
 
 export class PayPalNotificationHandler extends PayPalHandler {
-  private async handleMerchPayment(message: IpnMerchPmtType) {}
+  /**
+   * Handle merchant payment notification from PayPal
+   * and update Stripe invoice according to the payment_status
+   *
+   * @param message
+   */
+  private async handleMerchPayment(message: IpnMerchPmtType) {
+    const invoice = await this.stripeHelper.getInvoice(message.invoice);
+    if (!invoice) {
+      this.log.error('handleMerchPayment', {
+        message: 'Invoice not found',
+        ipnMessage: message,
+      });
+      throw error.internalValidationError('handleMerchPayment', {
+        message: 'Invoice not found',
+      });
+    }
+
+    if (invoice.status == null || !(['draft', 'open'].includes(invoice.status))) {
+      this.log.error('handleMerchPayment', {
+        message: 'Invoice not in draft or open status',
+        ipnMessage: message,
+      });
+      throw error.internalValidationError('handleMerchPayment', {
+        message: 'Invoice not in draft or open status',
+        transactionResponse: message.payment_status,
+        invoiceStatus: invoice.status,
+      });
+    }
+
+    switch (message.payment_status) {
+      case 'Completed':
+      case 'Processed':
+        return this.stripeHelper.payInvoiceOutOfBand(invoice);
+      case 'Pending':
+      case 'In-Progress':
+        return;
+      case 'Denied':
+      case 'Failed':
+      case 'Voided':
+      case 'Expired':
+        return this.stripeHelper.updatePaymentAttempts(invoice);
```

Comment:
I think we're going to need a more idempotent function here that takes into account the details of the transaction causing this. Because we update the payment attempts in `processInvoice` and we *may* need to update them here.